homeassistant/components/eq3btsmart/climate.py

```diff
@@ -0,0 +1,313 @@
+                if state is not None:
+                    try:
+                        return float(state.state)
+                    except ValueError:
+                        pass
+
+        return None
+
+    def _get_target_temperature(self) -> float | None:
+        """Return the target temperature."""
+
+        match self._eq3_config.target_temp_selector:
+            case TargetTemperatureSelector.TARGET:
+                return self._target_temperature
+            case TargetTemperatureSelector.LAST_REPORTED:
+                if self._thermostat.status is None:
+                    return None
+
+                return float(self._thermostat.status.target_temperature.value)
+
+    def _get_current_preset_mode(self) -> str:
+        """Return the current preset mode."""
+
+        if self._thermostat.status is None:
+            return PRESET_NONE
+        if self._thermostat.status.is_window_open:
+            return Preset.WINDOW_OPEN
+        if self._thermostat.status.is_boost:
+            return Preset.BOOST
+        if self._thermostat.status.is_low_battery:
+            return Preset.LOW_BATTERY
+        if self._thermostat.status.is_away:
+            return Preset.AWAY
+        if self._thermostat.status.operation_mode is OperationMode.ON:
+            return Preset.OPEN
+        if self._thermostat.status.presets is None:
+            return PRESET_NONE
+        if (
+            self._thermostat.status.target_temperature
+            == self._thermostat.status.presets.eco_temperature
+        ):
+            return Preset.ECO
+        if (
+            self._thermostat.status.target_temperature
+            == self._thermostat.status.presets.comfort_temperature
+        ):
+            return Preset.COMFORT
+
+        return PRESET_NONE
+
+    def _get_current_hvac_action(self) -> HVACAction:
+        """Return the current hvac action."""
+
+        if (
+            self._thermostat.status is None
+            or self._thermostat.status.operation_mode is OperationMode.OFF
+        ):
+            return HVACAction.OFF
+        if self._thermostat.status.valve == 0:
+            return HVACAction.IDLE
+        return HVACAction.HEATING
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
+        """Set new target temperature."""
+
+        if ATTR_HVAC_MODE in kwargs:
+            mode = kwargs.get(ATTR_HVAC_MODE)
+
+            if mode is None:
+                return
+
+            if mode == HVACMode.OFF:
+                raise ServiceValidationError(
+                    f"[{self._eq3_config.name}] Can't change HVAC mode to off while changing temperature",
+                )
+
+            await self.async_set_hvac_mode(mode)
+
+        temperature = kwargs.get(ATTR_TEMPERATURE)
+
+        if temperature is None:
+            return
+
+        previous_temperature = self._target_temperature
+        self._target_temperature = temperature
+
+        self.async_schedule_update_ha_state()
```

Comment:
Looks like more cases that can be replaced with `async_write_ha_state` in this function